### ISSUE TITLE
Regex not working as expected

### ISSUE THREAD

**Issue by shubh9194** [vespa-engine/vespa#24476]
**Describe the bug**
I am trying to use regex with attribute field in select query but not getting expected results.

Defined schema of fields used in query.
schema amDocument {

    document amDocument {
        field title type array<string> {
            indexing: summary | attribute | index
        }
       field collection type array<string> {
            indexing: summary | index
        }
      }
     field collection_attr type array<string> {
           indexing: input collection | attribute
           attribute: fast-search
      }
}

Document Ingested through postman
{
  "fields": {
 
   "table": "com.gs.document.store",
   "title": ["this is test document"],
   "collection":["public"],
    "sourcepath": [
      "2016-02.pdf"
    ],
    "filename": [
      "2016-02.pdf"
    ],
    "mimetype": [
      "application/pdf"
    ] 
  }
}

query used;
{
    "yql": "select * from sources * where collection_attr matches \"com\" ;",
    "timeout": "120s",
    "hits": 20
}
This should have returned no documents but it is returning the above ingested doc.

{
    "yql": "select * from sources * where title matches \"test document\" ;",
    "timeout": "120s",
    "hits": 20
}
this should have returned the document but it gives 0 documents
following query also not returning anything 
"yql": "select * from sources * where title matches \"test.*\" ;",

with title field, I am still able to get some results. like using title matches \"test\"
But with collection_attr, I get the document with almost any value in matches.

I am not able to understand why the above queries not working.
**Screenshots**

**Environment (please complete the following information):**
 - Infrastructure: Kubernetes
 - Versions [e.g. 22]

**Vespa version**
7.589

**Additional context**

**Comment by kkraune** [comment]
@geirst , I looked into the `matches` in an `array<string>` field - this works well:

````
schema music {
    document music {
        field album type array<string> {
            indexing: summary | attribute
        }
    }
}

{
    "put": "id:mynamespace:music::a-head-full-of-dreams",
    "fields": {
        "album": ["A Head Full of Dreams"]
    }
}

vespa query 'select * from sources * where album matches "notindoc"'
````

This gives 0 results.

However, adding `fast-search` makes anything match:
````
schema music {
    document music {
        field album type array<string> {
            indexing: summary | attribute
            attribute: fast-search
        }
    }
}

{
    "put": "id:mynamespace:music::a-head-full-of-dreams",
    "fields": {
        "album": ["A Head Full of Dreams"]
    }
}


$ vespa query 'select * from sources * where album matches "notindoc"'
{
    "root": {
        "id": "toplevel",
        "relevance": 1.0,
        "fields": {
            "totalCount": 1
        },
        "coverage": {
            "coverage": 100,
            "documents": 1,
            "full": true,
            "nodes": 1,
            "results": 1,
            "resultsFull": 1
        },
        "children": [
            {
                "id": "id:mynamespace:music::a-head-full-of-dreams",
                "relevance": 0.0017429193899782135,
                "source": "music",
                "fields": {
                    "sddocname": "music",
                    "documentid": "id:mynamespace:music::a-head-full-of-dreams",
                    "album": [
                        "A Head Full of Dreams"
                    ]
                }
            }
        ]
    }
}
````

**Comment by kkraune** [comment]
> Also from first comment, can we infer that regex(matches) does not work if field is defined both attribute and index as the query will be a text query (tokenized) in this case

yes - see https://docs.vespa.ai/en/schemas.html#indexing

@geirst is looking into the fast-search issue

**Comment by shubh9194** [comment]
@kkraune @geirst can you provide any ETA for fast-search issue

**Comment by kkraune** [comment]
@shubh9194  we have a process to evaluate tickets weekly, and this is not analyzed yet, so we cannot provide an ETA as of now

**Comment by bratseth** [comment]
(Triaged now)

**Comment by geirst** [comment]
This uncovered an edge case bug when using regex matching on a string attribute with fast-search. The bug could trigger if the prefix of the regex query term matched exactly one word in the underlying dictionary. If this word really didn't satisfy the regexp, all documents containing this word would be incorrectly returned.

In the example provided by Kristian, the regexp term "notindoc" has the prefix "". The empty prefix matches all words in the dictionary, but in this example the dictionary only contains a single word ("A Head Full of Dreams"). The bug then resulted in returning all documents containing "A Head Full of Dreams".

This has been fixed in Vespa 8.74.17.
